istioctl/cmd/sidecar.go

```diff
@@ -0,0 +1,361 @@
+}
+
+// Get and store the needed certificates
+// TODO: user internal generate-cert
+func createCertificates(kubeClient kube.ExtendedClient, dir string) error {
+	rootCert, err := kubeClient.CoreV1().ConfigMaps("istio-system").Get(context.Background(), "istio-ca-root-cert", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if err = ioutil.WriteFile(dir+"/root-cert.pem", []byte(rootCert.Data["root-cert.pem"]), tempPerms); err != nil {
+		return err
+	}
+
+	args := strings.Split("run istio.io/istio/security/tools/generate_cert -client -host spiffee://cluster.local/vm/vmname --out-priv key.pem --out-cert cert-chain.pem -mode citadel", " ")
+	cmd := exec.Command("go", args...)
+	cmd.Dir = dir
+	if err = cmd.Run(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.WorkloadGroup, clusterID, revision, dir string) error {
+	istioCM := "istio"
+	// Case with multiple control planes
+	if revision != "" {
+		istioCM = fmt.Sprintf("%s-%s", istioCM, revision)
+	}
+	istio, err := kubeClient.CoreV1().ConfigMaps("istio-system").Get(context.Background(), istioCM, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	temp, err := ioutil.TempFile("./", "")
+	defer os.Remove(temp.Name())
+	if _, err = temp.WriteString(istio.Data["mesh"]); err != nil {
+		return err
+	}
+	meshConfig, err := mesh.ReadMeshConfig(temp.Name())
+
+	labels := wg.Spec.Template.Metadata.Labels
+	// case where a user provided custom workload group has labels in the workload entry spec field
+	for k, v := range wg.Spec.Template.Spec.Labels {
+		labels[k] = v
+	}
+	we := wg.Spec.Template.Spec
+	md := meshConfig.DefaultConfig.ProxyMetadata
+	md["CANONICAL_SERVICE"], md["CANONICAL_REVISION"] = inject.ExtractCanonicalServiceLabels(labels, wg.Name)
+	md["DNS_AGENT"] = ""
+	md["POD_NAMESPACE"] = wg.Namespace
+	md["SERVICE_ACCOUNT"] = we.ServiceAccount
+	md["TRUST_DOMAIN"] = meshConfig.TrustDomain
+
+	md["ISTIO_META_CLUSTER_ID"] = clusterID
+	md["ISTIO_META_MESH_ID"] = string(meshConfig.DefaultConfig.MeshId)
+	md["ISTIO_META_NETWORK"] = we.Network
+	if portsJSON, err := json.Marshal(we.Ports); err == nil {
+		md["ISTIO_META_POD_PORTS"] = string(portsJSON)
+	}
+	md["ISTIO_META_WORKLOAD_NAME"] = wg.Name
+	labels["service.istio.io/canonical-name"] = md["CANONICAL_SERVICE"]
+	labels["service.istio.io/canonical-version"] = md["CANONICAL_REVISION"]
+	if labelsJSON, err := json.Marshal(labels); err == nil {
+		md["ISTIO_METAJSON_LABELS"] = string(labelsJSON)
+	}
+
+	meshYAML, err := yaml.Marshal(meshConfig)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(dir+"/mesh", meshYAML, tempPerms)
+}
+
+// Packs files inside source into target.tar
+func Tar(source, target string) error {
+	target = fmt.Sprintf("%s.tar", filepath.Base(target))
+	tarfile, err := os.Create(target)
+	if err != nil {
+		return err
+	}
+	defer tarfile.Close()
```

Comment:
this entire function could use some comments please. It also looks brittle as is.